Add docs for `vars.yml` (#8705)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/project-variables.md

pagination_next: "docs/build/environment-variables"
---

dbt provides a mechanism called [variables](/reference/dbt-jinja-functions/var) to provide data to models for compilation. Variables allow you to define configurable values for your project instead of hardcoding them in SQL.

<VersionBlock lastVersion="1.11">

You can define variables in two ways:

- In the [`dbt_project.yml`](/reference/dbt_project.yml) file
- On the command line

Variables defined in the `dbt_project.yml` act as project-wide defaults. These defaults apply anywhere the variable is referenced. You can override them at runtime using the `--vars` command-line argument, which is useful when you want to change a value for a specific run. For example, when testing with a different date range or running models with environment-specific settings.

</VersionBlock>

<VersionBlock firstVersion="1.12">

You can define variables in three ways:

> 1. In the `vars.yml` file (recommended for projects with multiple variables)

Reviewer: I would say it's recommended for projects with many variables or with variables they need to reference from _within_ `dbt_project.yml`.

> 1. In the `vars.yml` file (recommended for projects with several variables or for projects with variables you need to reference from within `dbt_project.yml`)

Reviewer 2: ```suggestion
1. In the `vars.yml` file (recommended for projects with several variables, or when you need to reference variables from within [`dbt_project.yml`](/reference/dbt_project.yml) since `vars.yml` is parsed first)
```

Author: @nataliefiann Committing this change instead to make this an overview since we discuss the order of precedence in other parts of the doc 
```suggestion
1. In the `vars.yml` file (recommended for projects with several variables, or when you need to reference variables from within [`dbt_project.yml`](/reference/dbt_project.yml)
```

2. In the `dbt_project.yml` file
3. On the command line

> Variables defined in the `vars.yml` or `dbt_project.yml` act as project-wide defaults. These defaults apply anywhere the variable is referenced. You can override them at runtime using the `--vars` command-line argument, which is useful when you want to change a value for a specific run. For example, when testing with a different date range or running models with environment-specific settings.

Reviewer 2: ```suggestion
Variables defined in the `vars.yml` or in `dbt_project.yml` act as project-wide defaults. These defaults apply anywhere the variable is referenced. You can override them at runtime using the `--vars` command-line argument, which is useful when you want to change a value for a specific run. For example, when testing with a different date range or running models with environment-specific settings.
```

Reviewer 2: Hiya Bianca,

Just added the word "in" to emphasise variables can only be defined in one or the other (with ref to the error mentioned in the Githib issue that mentions the project will throw error if variables are set in both the `vars.yml` and `dbt_project.yml`

</VersionBlock>

You might use variables to [configure timezones](https://github.com/dbt-labs/snowplow/blob/0.3.9/dbt_project.yml#L22), set reporting date ranges, [avoid hardcoding table names](https://github.com/dbt-labs/quickbooks/blob/v0.1.0/dbt_project.yml#L23), or otherwise control how models are compiled.

To use a variable in a model, hook, or macro, use the `{{ var('...') }}` function. The `var()` function retrieves the value defined in your project or passed using `--vars`. For more information, see [About var function](/reference/dbt-jinja-functions/var).

Note, refer to [YAML tips](/docs/build/dbt-tips#yaml-tips) for more YAML information.

<VersionBlock firstVersion="1.12">

### Defining variables in `vars.yml`<Lifecycle status="beta" />

> :::info Beta feature

Reviewer: @eliasdefaria does this feel good as a callout for "this is in beta in core" (i.e. doesn't yet work for fusion, but will by GA)

The `vars.yml` file is a beta feature in <Constant name="core" /> v1.12.
:::

> `vars.yml` is a dedicated file for project variables located at the project root. Because dbt parses it _before_ `dbt_project.yml`, you can [reference variables from `vars.yml` in `dbt_project.yml`](#using-varsyml-variables-in-dbt_projectyml) using `{{ var('...') }}`. 

Reviewer 2: ```suggestion
`vars.yml` is a dedicated file for project variables located at the project root. Because dbt parses it _before_ `dbt_project.yml`, variables from `vars.yml` are available when `dbt_project.yml` is evaluated, so you can reference them in `dbt_project.yml` using `{{ var('...') }}`. For more information, refer to [Using vars.yml variables in dbt_project.yml](#using-varsyml-variables-in-dbt_projectyml).
```

> To define variables in a `vars.yml` file, add a `vars` config. These variables can be scoped globally, or to a specific package imported in your project. See the following example of variables in a `vars.yml` file:

Author: ```suggestion
To define variables in a `vars.yml` file, add a `vars` config. You can scope these variables globally, or to a specific package imported in your project. See the following example of variables in a `vars.yml` file:
```

> To define variables in a `vars.yml` file, add a `vars` config. You can scope these variables globally, or to a specific package imported in your project. See the following example of variables in a `vars.yml` file:

Reviewer 2: ```suggestion
To define variables in a `vars.yml` file, add a `vars` config. You can scope these variables globally, or to a specific package imported in your project. The following example shows variables in a `vars.yml` file:
```

Reviewer 2: Hiya Bianca

I've amended this to adhere to the accessibility language.

<File name='vars.yml'>

```yaml
vars:
  # The `start_date` variable is accessible to all resources
  start_date: '2016-06-01'
  

>   # The `platforms` variable is only accessible to resources in the my_dbt_project project

Reviewer: is this accurate? @sriramr98 can you confirm?

Reviewer 3: Yes this is accurate. All existing behaviour of variables declared in `dbt_project.yml` is preserved when using `vars.yml`

  my_dbt_project:
    platforms: ['web', 'mobile']
  
  # The `app_ids` variable is only accessible to resources in the snowplow package
  snowplow:
    app_ids: ['marketing', 'app', 'landing-page']
```

</File>

> You cannot define variables in both `vars.yml` and `dbt_project.yml`; you can only use one or the other. If both files contain a `vars` block with definitions, dbt raises an error.

Reviewer 2: Hiya Bianca,

It might be worth rewording the sentence as the first part confirms the variables can't be used at the same time. Maybe something like: "You cannot define variables in both `vars.yml` and `dbt_project.yml`. If both files contain a vars block with definitions, dbt raises an error? (I'll add as a suggestion)

> You cannot define variables in both `vars.yml` and `dbt_project.yml`; you can only use one or the other. If both files contain a `vars` block with definitions, dbt raises an error.

Reviewer 2: ```suggestion
You cannot define variables in both `vars.yml` and `dbt_project.yml`. If both files contain a `vars` block with definitions, dbt raises an error.
```

> You cannot define variables in both `vars.yml` and `dbt_project.yml`. If both files contain a `vars` block with definitions, dbt raises an error.

Reviewer 2: ```suggestion
You cannot define variables in both `vars.yml` and `dbt_project.yml`; you can only use one or the other. If both files contain a `vars` block with definitions, dbt raises an error.
```

Reviewer 2: Hiya @luna-bianca 

If you decide to go with this suggestion, this could be a reusable snippet

Author: Hey @nataliefiann , this is the same as my original draft, and I removed "; you can only use one or the other" based on your feedback [here](https://github.com/dbt-labs/docs.getdbt.com/pull/8705#discussion_r2890698789). Happy to revert to the original draft -- just want to make sure we're aligned!

> If `vars.yml` is empty or doesn't exist, variables defined in `dbt_project.yml` are used instead. If `vars.yml` exists but has no top-level `vars` key, variables are resolved from the command line or `dbt_project.yml`.

Reviewer: ```suggestion
If `vars.yml` is empty or doesn't exist, variables defined in `dbt_project.yml` are used instead. 
```

I don't think we need a callout for this, since we already say you need to nest your variables under `vars:`

#### Using `vars.yml` variables in `dbt_project.yml`<Lifecycle status="beta" />

> Because `vars.yml` is parsed before `dbt_project.yml`, you can reference variables from `vars.yml` within `dbt_project.yml` using `{{ var('...') }}`. Variables referenced in `dbt_project.yml` are resolved only from `vars.yml` or the `--vars` command-line argument, not from the `vars` block in `dbt_project.yml` itself. If `dbt_project.yml` expects a variable that is defined only in its own `vars` block, dbt raises an error.

Reviewer 2: ```suggestion
Because `vars.yml` is parsed before `dbt_project.yml`, you can reference variables from `vars.yml` within `dbt_project.yml` using `{{ var('...') }}`. Variables referenced in `dbt_project.yml` are resolved only from `vars.yml` or the `--vars` command-line argument, not from the `vars` block in `dbt_project.yml` itself. If `dbt_project.yml` references a variable that is defined only in its own `vars` block, dbt raises an error.
```

</File>

<VersionBlock firstVersion="1.12">

> You cannot define variables in both `vars.yml` and `dbt_project.yml`. If both files contain a `vars` block with definitions, dbt raises an error.

Reviewer 2: ```suggestion
You cannot define variables in both `vars.yml` and `dbt_project.yml`; you can only use one or the other. If both files contain a `vars` block with definitions, dbt raises an error.
```

Author: Same feedback [here](https://github.com/dbt-labs/docs.getdbt.com/pull/8705#discussion_r2905036696)

> You cannot define variables in both `vars.yml` and `dbt_project.yml`; you can only use one or the other. If both files contain a `vars` block with definitions, dbt raises an error.

Reviewer 2: ```suggestion
You cannot define variables in both `vars.yml` and `dbt_project.yml`. If both files contain a `vars` block with definitions, dbt raises an error.
```

Reviewer 2: Same as earlier suggestion

### Variable precedence

<VersionBlock lastVersion="1.11">

Variables defined with the `--vars` command line argument override variables defined in the `dbt_project.yml` file. They are globally scoped and accessible to the root project and all installed packages.

The order of precedence for variable declaration is as follows (highest priority first):

1. The variables defined on the command line with `--vars`
2. The package-scoped variable declaration in the root `dbt_project.yml` file
3. The global variable declaration in the root `dbt_project.yml` file
4. If this node is defined in a package: variable declarations in that package's `dbt_project.yml` file
5. The variable's default argument (if one is provided)

</VersionBlock>

<VersionBlock firstVersion="1.12">

Variables defined with the `--vars` command line argument override variables defined in `vars.yml` or `dbt_project.yml`. They are globally scoped and accessible to the root project and all installed packages.

The order of precedence for variable declaration is as follows (highest priority first):

1. The variables defined on the command line with `--vars`

> 2. The package-scoped variable declaration in the root `vars.yml` file (if present)

Reviewer: again, i wasn't aware we had package-scoped variable declaration. we should confirm this is accurate

Reviewer: oh - i do see that in existing [docs](https://docs.getdbt.com/docs/build/project-variables)
